include/seqan3/alignment/matrix/detail/combined_score_and_trace_matrix.hpp

```diff
@@ -0,0 +1,354 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::combined_score_and_trace_matrix.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/std/ranges>
+
+#include <seqan3/alignment/matrix/detail/affine_cell_proxy.hpp>
+#include <seqan3/alignment/matrix/detail/matrix_coordinate.hpp>
+#include <seqan3/core/concept/core_language.hpp>
+#include <seqan3/core/type_traits/range.hpp>
+#include <seqan3/range/views/zip.hpp>
+
+namespace seqan3::detail
+{
+/*!\brief An alignment matrix that combines a score matrix and a trace matrix to a common interface.
+ * \ingroup alignment_matrix
+ * \implements std::ranges::input_range
+ *
+ * \tparam score_matrix_t The type of the score matrix; must model std::ranges::input_range (for more requirements see
+ *                        details section).
+ * \tparam trace_matrix_t The type of the trace matrix; must model std::ranges::input_range (for more requirements see
+ *                        details section).
+ *
+ * \details
+ *
+ * To compute the actual alignment an additional trace matrix is required, which stores the trace directions
+ * to the previous cells for any computed cell within the alignment matrix. This matrix combines a given score and
+ * trace matrix to a complete alignment matrix. The iterator increments over the columns of each underlying matrix and
+ * combines them into a zip view, i.e. the combined alignment matrix is a range over ranges with the inner range also
+ * modelling std::ranges::input_range.
+ * To provide a uniform interface to the alignment algorithm, the returned zip view is further transformed, such that
+ * its reference type is a seqan3::detail::affine_cell_proxy.
+ */
+template <std::ranges::input_range score_matrix_t, std::ranges::input_range trace_matrix_t>
+//!\cond
+    requires (std::ranges::input_range<std::ranges::range_reference_t<score_matrix_t>> &&
+              std::ranges::input_range<std::ranges::range_reference_t<trace_matrix_t>>)
+//!\endcond
+class combined_score_and_trace_matrix
+{
+private:
+    //!\brief The score type extracted from the score matrix. (TODO: make this less complicated by exposing the score type).
```

Comment:
Line length.
Did you want to do this in this PR or did you want to create a backlog item?

Reply:
Yes, I want to do it later, since it means changing some other files as well, that would not fit into this PR.